pylearn2/linear/cudnn2d.py

```diff
@@ -0,0 +1,195 @@
+                 batch_size,
+                 input_space,
+                 output_axes=('b', 'c', 0, 1),
+                 subsample=(1, 1),
+                 border_mode='valid',
+                 filters_shape=None,
+                 message=''):
+
+        assert batch_size is None or batch_size > 0
+        self._input_space = input_space
+        self._output_axes = output_axes
+        self._subsample = tuple(subsample)
+        self._filters = filters
+
+        super(Cudnn2D, self).__init__(
+            filters=filters,
+            img_shape=(batch_size, input_space.num_channels,
+                       input_space.shape[0], input_space.shape[1]),
+            subsample=self.subsample,
+            border_mode=border_mode,
+            filters_shape=filters.get_value(borrow=True).shape,
+            message=message
+        )
+
+        # conv_op has to be changed
+        self._conv_op = GpuDnnConv()
+        self._desc = GpuDnnConvDesc(border_mode=border_mode,
+                                    subsample=self.subsample,
+                                    conv_mode='conv')
+
+    @functools.wraps(P2LT.get_params)
+    def get_params(self):
+        return [self._filters]
+
+    @functools.wraps(P2LT.get_weights_topo)
+    def get_weights_topo(self, borrow):
+        """
+        Parameters
+        ----------
+        borrow : TODO
+            TODO
+        """
+        return np.transpose(self._filters.get_value(borrow=borrow),
+                            (0, 2, 3, 1))
+
+    def lmul(self, x):
+        """
+        .. todo::
+
+            WRITEME properly
+
+        dot(x, A)
+
+        This method overrides the original Conv2D lmul to make it work
+        with arbitrary axis orders
+
+        Parameters
+        ----------
+        x : TODO
+            TODO
+        """
+
+        # x must be formatted as batch index, channel, topo dim 0, topo dim 1
+        # for use with conv2d, so check what the current input space format is
+        assert x.ndim == 4
+        axes = self._input_space.axes
+        assert len(axes) == 4
+
+        op_axes = ('b', 'c', 0, 1)
+
+        if tuple(axes) != op_axes:
+            x = x.dimshuffle(*[axes.index(ax) for ax in op_axes])
+
+        # The calling format has to be changed
+        img = gpu_contiguous(x)
+        kerns = gpu_contiguous(self._filters)
+        desc = self._desc(img.shape, kerns.shape)
+        rval = self._conv_op(img, kerns, desc)
+
+        # Format the output based on the output space
+        axes = self._output_axes
+        assert len(axes) == 4
+
+        if tuple(axes) != op_axes:
+            rval = rval.dimshuffle(*[axes.index(ax) for ax in op_axes])
+
+        return rval
+
+    def set_batch_size(self, batch_size):
+        """
+        .. todo::
+
+            WRITEME
+
+        Parameters
+        ----------
+        batch_size : TODO
+            TODO
+        """
+        self._img_shape = tuple([batch_size] + list(self._img_shape[1:]))
```

Comment:
`self._img_shape` contains the shape for the whole input tensor (bc01).
There is no `self._batch_size`.